### DISCUSSION TITLE
prisma execute a function while running a prisma query

### DISCUSSION THREAD

**Discussion by md186** [prisma/prisma#14554] [Q&A]
Hi,

I have a query where I'm searching for users matching some criteria. Now every user has a location with coordinates (Longitude and latitude) and I need to compare the distance between the coordinates of the query with a given coordinate I have. This is my query which does search for the matching criteria only.

```
import prisma from "../../../utils/prisma";
import { allowPost } from "../auth/middlewares";

export default allowPost(async function handler(req, res) {
const data = req.body;
const { userId, searchAgeMin, searchAgeMax, gender, country, longitude, latitude, allowedDistance } = data;

  const id = parseInt(userId)
  const ageMin = parseInt(searchAgeMin)
  const ageMax = parseInt(searchAgeMax)

  const matchingPartner = await prisma.user.findMany({
        where: {
          AND: [
            { dislikedBy: { none: { userId: id } },},
            { likedBy: { none: { userId: id } },},

            { profile: { age: { gte: ageMin } }, },
            { profile: { age: { lte: ageMax } }, },

            { profile: { gender: { equals: gender } }, },
            { profile: { country: { equals: country } }, },

           // I need to get the coordinate of the user from my database here, pass it to a function on my api and compare the result 
          // with a given coordinate
          ],
      },
      select: {
        userId: true,
        firstName: true,
        lastName: true,
        profile: {
          select: {
            age: true,
            gender: true,
            country: true,
          },
        },
      },
      take: 5,
  });

  res.json({ data: matchingPartner });
});

```
This is my example data im providing the api. There is a longitude and latitude of a give user which I have to compare to all of the users coordinate in my query within a radius for example 5 KM:

```
{
  "userId": "11",
  "searchAgeMin": "30",
  "searchAgeMax": "30",
  "gender": "male",
  "country": "usa",
  "longitude": "51.5175",
  "latitude": "7.4678",
   "allowedDistance":"5000"
}
```

Let's say im searching users with all the criteria and im searching for users where theire coordinate has to be less than 5000m from the coordinate im providing to the api. With the current setup I would use the above query for 5 users matching the criteria and after that I have to check the distance separately in another function, let's call it `getDistanceOfCoordinates()`. When I have 10.000 Users, I would need to run the above query many times until I have 5 matching results. Im wondering if there is a nice way to run my `getDistanceOfCoordinates()` while running the query, so I don't end up in 5000 query runs.

Does somebody have an idea how to include my comparison together with the prisma query?

**Comment by austincrim** [comment]
Hey @md186 👋,

Could you share some details about your schema and how you are calculating the distance from coordinates?

It seems like the best bet is to find a way to filter by distance/coordinates in your query's `where` condition.

**Comment by md186** [comment]
Hi @austincrim ,

thanks for your reply. Sure, this is my simplified Schema:

```
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
  shadowDatabaseUrl = env("SHADOW_DATABASE_URL")
}

model User {
  userId        Int         @id @default(autoincrement())
  registeredAt  DateTime   @default(now())
  firstName     String      @db.VarChar(250)
  lastName      String      @db.VarChar(250)
  email         String      @db.VarChar(250)
  password      String      @db.VarChar(250)
  socketId      String?      @db.VarChar(250)
  role          String     @default("user")
  confirmedAccount  DateTime?
  contacts User[]  @relation("ContactsList")
  addedByContacts  User[]  @relation("ContactsList")
  conversation  Conversation[]
  likedBy User[]  @relation("LikeDislikeList")
  dislikedBy  User[]  @relation("LikeDislikeList")
  profile Profile?
}

model Profile {
  profId        Int         @id @default(autoincrement())
  longitude       String
  latidude       String
  age Int
  gender String
  country String
  user   User @relation(fields: [uId], references: [userId])
  uId Int @unique
}
```

I have another api function which takes two longitudes and latitudes and tells me if the distance of my users location is in a radius X compared to the other coordinate. The api just tells me if it is in the given radius by saying true or false, but I can ofcourse rewrite it. I really have no idea how to perform my function for the comparison between the coordinates while im running the query for the users. Anyways this it the comparison api function: (for example: checks if 51.525/7.4575 is within a radius of 5 km from 51.5175/7.4678)

```
import { allowPost } from "../auth/middlewares";
import { isPointWithinRadius } from 'geolib';

export default allowPost(async function handler(req, res) {
const data = req.body;
const { startLatitude, startLongitude, endLatitude, endLongitude, radius } = data;

const sLatitude = parseFloat(startLatitude)
const sLongitude = parseFloat(startLongitude)
const eLatitude = parseFloat(endLatitude)
const eLongitude = parseFloat(endLongitude)
const sRadius = parseInt(radius)

try {
const isInRadius = isPointWithinRadius(
    { latitude: sLatitude, longitude: sLongitude },
    { latitude: eLatitude, longitude: eLongitude },
    sRadius
);
res.status(201).json(isInRadius);
} catch (error) {
    res.status(422).json({ message: error });
    return;
}
});
```

Thanks in advance!

**Reply by md186** [reply]
I Also updated my question post with a data json im providing to my api, some more explanation and I updated the query itself to my latest code state. I always have the longitude and latitude and an allowed radius. i need to compare the coordinates with the coordinates of my users from the query, comparing them with the allowed radius and get the allowed users back as query result. Im excited for your approach!

**Reply by austincrim** [reply]
Ah, that helps, thank you!

The only way I can see to accomplish what you want is to move the distance-calculating logic into the database. I see you are using Postgres, you could take a look at the [PostGIS extension](https://postgis.net/). Then you could query for users and filter on distance in the same query.

As a note, [Prisma doesn't fully support PostGIS](https://github.com/prisma/prisma/issues/2789), so you would also need to use `$queryRaw`.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
